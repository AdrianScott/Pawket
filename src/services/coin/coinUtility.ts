import { Hex0x, prefix0x } from "./condition";
import { Bytes, bigint_to_bytes } from "clvm";
import { GetPuzzleApiCallback } from "../transfer/transfer";
import { CoinItem } from "@/models/wallet";
import { CoinSpend, OriginCoin } from "../spendbundle";
import store from "@/store";

export interface NetworkContext {
  prefix: string;
  symbol: string;
  chainId: string;
  api: GetPuzzleApiCallback;
}

export interface NetworkContextWithOptionalApi {
  prefix: string;
  symbol: string;
  chainId: string;
  api?: GetPuzzleApiCallback;
}

export interface CompatibleCoin {
  amount: bigint | number;
  parent_coin_info: Hex0x | string;
  puzzle_hash: Hex0x | string;
}

export interface LockedCoin {
  coinName: string;
  coin: OriginCoin;
  transactionTime: number;
  network: string;
  symbol?: string;
  accountFinger: number;
}

export interface PendingTransaction {
  coin: LockedCoin[];
  time: number;
  network: string;
  amount: { [key: string]: bigint };
}

export async function lockCoins(coinSpends: CoinSpend[], transactionTime: number, network: string): Promise<void> {
  const lcStr = localStorage.getItem("LOCKED_COINS");
  const lc: LockedCoin[] = lcStr ? JSON.parse(lcStr) : [];
  const accountFinger = store.state.account.accounts[store.state.account.selectedAccount].key.fingerprint;
  for (const cs of coinSpends) {
    const coinName = getCoinName(cs.coin);
    lc.push({ coinName: coinName, coin: cs.coin, transactionTime: transactionTime, network: network, symbol: cs.coin.symbol, accountFinger: accountFinger });
  }
  localStorage.setItem("LOCKED_COINS", JSON.stringify(lc));
}

export async function unlockCoins(coins: OriginCoin[]): Promise<void> {
  const lcStr = localStorage.getItem("LOCKED_COINS");
  let lc: LockedCoin[] = lcStr ? JSON.parse(lcStr) : [];
  for (const coin of coins) {
    const cname = getCoinName(coin);
    lc = lc.filter(c => c.coinName !== cname);
  }
  localStorage.setItem("LOCKED_COINS", JSON.stringify(lc));
}

export function coinFilter(coins: OriginCoin[], network: string): OriginCoin[] {
  const lcStr = localStorage.getItem("LOCKED_COINS");
  let lc: LockedCoin[] = lcStr ? JSON.parse(lcStr) : [];
  const accountFinger = store.state.account.accounts[store.state.account.selectedAccount].key.fingerprint;
  lc = lc.filter((l) => l.network == network && l.accountFinger == accountFinger);
  return coins.filter(coin => {
    const name = getCoinName(coin);
    return lc.findIndex(c => c.coinName == name) == -1;
  });
}

export function getCoinName0x(coin: CompatibleCoin): Hex0x {
  return prefix0x(getCoinNameHex(coin).hex());
}

export function getCoinName(coin: CompatibleCoin): string {
  return getCoinNameHex(coin).hex();
}

export function getCoinNameHex(coin: CompatibleCoin): Bytes {
  const a = bigint_to_bytes(BigInt(coin.amount), { signed: true });
  const pci = Bytes.from(coin.parent_coin_info, "hex");
  const ph = Bytes.from(coin.puzzle_hash, "hex");
  const cont = pci.concat(ph).concat(a);
  const coinname = Bytes.SHA256(cont);
  return coinname;
}

export function convertToOriginCoin(coin: CoinItem | { amount: number, parent_coin_info: Hex0x | string, puzzle_hash: Hex0x | string, symbol?: string }): OriginCoin {
  return "parentCoinInfo" in coin
    ? {
      amount: BigInt(coin.amount),
      parent_coin_info: coin.parentCoinInfo,
      puzzle_hash: coin.puzzleHash,
      symbol: coin.symbol,
    }
    : {
      amount: BigInt(coin.amount),
      parent_coin_info: prefix0x(coin.parent_coin_info),
      puzzle_hash: prefix0x(coin.puzzle_hash),
      symbol: coin.symbol
    };
}