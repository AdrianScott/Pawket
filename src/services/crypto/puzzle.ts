import * as clvm_tools from "clvm_tools";
import { bech32m } from "@scure/base";
import { Bytes } from "clvm";
import { PrivateKey } from "@chiamine/bls-signatures";
import utility from "./utility";
import { assemble, disassemble } from "clvm_tools/clvm_tools/binutils";
import { Instance } from "../util/instance";
import { modsdict } from "../coin/mods";
import { unprefix0x } from "../coin/condition";

export interface ExecuteResultCondition {
  op: number;
  args: string[];
}

export interface ExecuteResult {
  raw: string;
  conditions: ExecuteResultCondition[];
}

export type AddressType = "Observed" | "Hardened" | "Unknown";

export interface PuzzleDetail extends PuzzleAddress {
  privateKey: PrivateKey;
  puzzle: string;
}

export interface PuzzleAddress {
  hash: string;
  address: string;
  type?: AddressType;
}

export type ConditionArgs = (Uint8Array | undefined | ConditionArgs[]);

export interface ConditionEntity {
  code: number;
  args: ConditionArgs[];
}

export const catClvmTreehash = "0x72dec062874cd4d3aab892a0906688a1ae412b0109982e1797a170add88bdcdc";

class PuzzleMaker {

  public getPuzzle(synPubkey: string): string {
    const puzzle = `(a (q 2 (q 2 (i 11 (q 2 (i (= 5 (point_add 11 (pubkey_for_exp (sha256 11 (a 6 (c 2 (c 23 ()))))))) (q 2 23 47) (q 8)) 1) (q 4 (c 4 (c 5 (c (a 6 (c 2 (c 23 ()))) ()))) (a 23 47))) 1) (c (q 50 2 (i (l 5) (q 11 (q . 2) (a 6 (c 2 (c 9 ()))) (a 6 (c 2 (c 13 ())))) (q 11 (q . 1) 5)) 1) 1)) (c (q . ${synPubkey}) 1))`;
    return puzzle;
  }

  public async getSyntheticKey(pubkey: string): Promise<string> {
    if (!pubkey.startsWith("0x")) pubkey = "0x" + pubkey;
    const hidden_puzzle_hash = "0x711d6c4e32c92e53179b199484cf8c897542bc57f2b22582799f9d657eec4699";

    let output: string[] = [];
    clvm_tools.setPrintFunction((...args) => output = args)

    clvm_tools.go(
      "brun",
      "(point_add 2 (pubkey_for_exp (sha256 2 5)))",
      `(${pubkey} ${hidden_puzzle_hash})`
    );
    const synPubkey = output[0];

    return synPubkey;
  }

  public async getPuzzleHash(pubkey: string): Promise<string> {
    const synPubkey = await this.getSyntheticKey(pubkey);
    const puzzle = this.getPuzzle(synPubkey);
    return await this.getPuzzleHashFromPuzzle(puzzle);
  }

  public async getPuzzleHashFromPuzzle(puzzle: string): Promise<string> {
    let output: string[] = [];
    clvm_tools.setPrintFunction((...args) => output = args)

    clvm_tools.go("opc", "-H", puzzle);
    const puzzleHash = output[0];

    return puzzleHash;
  }

  public async encodePuzzle(puzzle: string): Promise<string> {
    let output: string[] = [];
    clvm_tools.setPrintFunction((...args) => output = args)

    clvm_tools.go("opc", puzzle);
    const encodedPuzzle = output[0];

    return encodedPuzzle;
  }

  public async disassemblePuzzle(puzzle_hex: string): Promise<string> {
    puzzle_hex = unprefix0x(puzzle_hex);
    let output: string[] = [];
    clvm_tools.setPrintFunction((...args) => output = args)

    clvm_tools.go("opd", puzzle_hex);
    const puzzle = output[0];

    return puzzle;
  }

  public getCatPuzzle(synPubkey: string, assetId: string): string {
    if (!synPubkey.startsWith("0x")) synPubkey = "0x" + synPubkey;
    if (!assetId.startsWith("0x")) assetId = "0x" + assetId;

    const puzzle = `(a (q 2 (q 2 94 (c 2 (c (c 5 (c (sha256 44 5) (c 11 ()))) (c (a 23 47) (c 95 (c (a 46 (c 2 (c 23 ()))) (c (sha256 639 1407 2943) (c -65 (c 383 (c 767 (c 1535 (c 3071 ())))))))))))) (c (q (((-54 . 61) 70 2 . 51) (60 . 4) 1 1 . -53) ((a 2 (i 5 (q 2 50 (c 2 (c 13 (c (sha256 34 (sha256 44 52) (sha256 34 (sha256 34 (sha256 44 92) 9) (sha256 34 11 (sha256 44 ())))) ())))) (q . 11)) 1) (a (i 11 (q 2 (i (= (a 46 (c 2 (c 19 ()))) 2975) (q 2 38 (c 2 (c (a 19 (c 95 (c 23 (c 47 (c -65 (c 383 (c 27 ()))))))) (c 383 ())))) (q 8)) 1) (q 2 (i 23 (q 2 (i (not -65) (q . 383) (q 8)) 1) (q 8)) 1)) 1) (c (c 5 39) (c (+ 11 87) 119)) 2 (i 5 (q 2 (i (= (a (i (= 17 120) (q . 89) ()) 1) (q . -113)) (q 2 122 (c 2 (c 13 (c 11 (c (c -71 377) ()))))) (q 2 90 (c 2 (c (a (i (= 17 120) (q 4 120 (c (a 54 (c 2 (c 19 (c 41 (c (sha256 44 91) (c 43 ())))))) 57)) (q 2 (i (= 17 36) (q 4 36 (c (sha256 32 41) 57)) (q . 9)) 1)) 1) (c (a (i (= 17 120) (q . 89) ()) 1) (c (a 122 (c 2 (c 13 (c 11 (c 23 ()))))) ())))))) 1) (q 4 () (c () 23))) 1) ((a (i 5 (q 4 9 (a 38 (c 2 (c 13 (c 11 ()))))) (q . 11)) 1) 11 34 (sha256 44 88) (sha256 34 (sha256 34 (sha256 44 92) 5) (sha256 34 (a 50 (c 2 (c 7 (c (sha256 44 44) ())))) (sha256 44 ())))) (a (i (l 5) (q 11 (q . 2) (a 46 (c 2 (c 9 ()))) (a 46 (c 2 (c 13 ())))) (q 11 44 5)) 1) (c (c 40 (c 95 ())) (a 126 (c 2 (c (c (c 47 5) (c 95 383)) (c (a 122 (c 2 (c 11 (c 5 (q ()))))) (c 23 (c -65 (c 383 (c (sha256 1279 (a 54 (c 2 (c 9 (c 2815 (c (sha256 44 45) (c 21 ())))))) 5887) (c 1535 (c 3071 ()))))))))))) 2 42 (c 2 (c 95 (c 59 (c (a (i 23 (q 9 45 (sha256 39 (a 54 (c 2 (c 41 (c 87 (c (sha256 44 -71) (c 89 ())))))) -73)) ()) 1) (c 23 (c 5 (c 767 (c (c (c 36 (c (sha256 124 47 383) ())) (c (c 48 (c (sha256 -65 (sha256 124 21 (+ 383 (- 735 43) 767))) ())) 19)) ()))))))))) 1)) (c (q . ${catClvmTreehash}) (c (q . ${assetId}) (c (q 2 (q 2 (q 2 (i 11 (q 2 (i (= 5 (point_add 11 (pubkey_for_exp (sha256 11 (a 6 (c 2 (c 23 ()))))))) (q 2 23 47) (q 8)) 1) (q 4 (c 4 (c 5 (c (a 6 (c 2 (c 23 ()))) ()))) (a 23 47))) 1) (c (q 50 2 (i (l 5) (q 11 (q . 2) (a 6 (c 2 (c 9 ()))) (a 6 (c 2 (c 13 ())))) (q 11 (q . 1) 5)) 1) 1)) (c (q . ${synPubkey}) 1)) 1))))`;

    return puzzle;
  }

  public async getCatPuzzleHash(pubkey: string, assetId: string): Promise<string> {
    const synPubkey = await this.getSyntheticKey(pubkey);
    const puzzle = this.getCatPuzzle(synPubkey, assetId);

    return await this.getPuzzleHashFromPuzzle(puzzle);
  }

  public getCatSettlementPuzzle(assetId: string): string {
    if (!assetId.startsWith("0x")) assetId = "0x" + assetId;
    return `(a (q 2 (q 2 94 (c 2 (c (c 5 (c (sha256 44 5) (c 11 ()))) (c (a 23 47) (c 95 (c (a 46 (c 2 (c 23 ()))) (c (sha256 639 1407 2943) (c -65 (c 383 (c 767 (c 1535 (c 3071 ())))))))))))) (c (q (((-54 . 61) 70 2 . 51) (60 . 4) 1 1 . -53) ((a 2 (i 5 (q 2 50 (c 2 (c 13 (c (sha256 34 (sha256 44 52) (sha256 34 (sha256 34 (sha256 44 92) 9) (sha256 34 11 (sha256 44 ())))) ())))) (q . 11)) 1) (a (i 11 (q 2 (i (= (a 46 (c 2 (c 19 ()))) 2975) (q 2 38 (c 2 (c (a 19 (c 95 (c 23 (c 47 (c -65 (c 383 (c 27 ()))))))) (c 383 ())))) (q 8)) 1) (q 2 (i 23 (q 2 (i (not -65) (q . 383) (q 8)) 1) (q 8)) 1)) 1) (c (c 5 39) (c (+ 11 87) 119)) 2 (i 5 (q 2 (i (= (a (i (= 17 120) (q . 89) ()) 1) (q . -113)) (q 2 122 (c 2 (c 13 (c 11 (c (c -71 377) ()))))) (q 2 90 (c 2 (c (a (i (= 17 120) (q 4 120 (c (a 54 (c 2 (c 19 (c 41 (c (sha256 44 91) (c 43 ())))))) 57)) (q 2 (i (= 17 36) (q 4 36 (c (sha256 32 41) 57)) (q . 9)) 1)) 1) (c (a (i (= 17 120) (q . 89) ()) 1) (c (a 122 (c 2 (c 13 (c 11 (c 23 ()))))) ())))))) 1) (q 4 () (c () 23))) 1) ((a (i 5 (q 4 9 (a 38 (c 2 (c 13 (c 11 ()))))) (q . 11)) 1) 11 34 (sha256 44 88) (sha256 34 (sha256 34 (sha256 44 92) 5) (sha256 34 (a 50 (c 2 (c 7 (c (sha256 44 44) ())))) (sha256 44 ())))) (a (i (l 5) (q 11 (q . 2) (a 46 (c 2 (c 9 ()))) (a 46 (c 2 (c 13 ())))) (q 11 44 5)) 1) (c (c 40 (c 95 ())) (a 126 (c 2 (c (c (c 47 5) (c 95 383)) (c (a 122 (c 2 (c 11 (c 5 (q ()))))) (c 23 (c -65 (c 383 (c (sha256 1279 (a 54 (c 2 (c 9 (c 2815 (c (sha256 44 45) (c 21 ())))))) 5887) (c 1535 (c 3071 ()))))))))))) 2 42 (c 2 (c 95 (c 59 (c (a (i 23 (q 9 45 (sha256 39 (a 54 (c 2 (c 41 (c 87 (c (sha256 44 -71) (c 89 ())))))) -73)) ()) 1) (c 23 (c 5 (c 767 (c (c (c 36 (c (sha256 124 47 383) ())) (c (c 48 (c (sha256 -65 (sha256 124 21 (+ 383 (- 735 43) 767))) ())) 19)) ()))))))))) 1)) (c (q . ${catClvmTreehash}) (c (q . ${assetId}) (c (q 2 (q 2 10 (c 2 (c 3 ()))) (c (q (51 . 62) (a (i 5 (q 4 (c 12 (c (a 30 (c 2 (c 9 ()))) ())) (a 22 (c 2 (c 25 (c (a 10 (c 2 (c 13 ()))) ()))))) ()) 1) (a (i 5 (q 4 (c 8 9) (a 22 (c 2 (c 13 (c 11 ()))))) (q . 11)) 1) 2 (i (l 5) (q 11 (q . 2) (a 30 (c 2 (c 9 ()))) (a 30 (c 2 (c 13 ())))) (q 11 (q . 1) 5)) 1) 1)) 1))))`;
  }

  public async getAddress(pubkey: string, prefix: string): Promise<string> {
    const puzzleHash = await this.getPuzzleHash(pubkey);

    const address = bech32m.encode(prefix, bech32m.toWords(utility.fromHexString(puzzleHash)));
    // console.log(address);
    return address;
  }

  public getAddressesFromPuzzleHash(puzzleHashes: string[], prefix: string): string[] {
    const arr = [];
    for (let i = 0; i < puzzleHashes.length; i++) {
      const h = puzzleHashes[i];
      arr.push(this.getAddressFromPuzzleHash(h, prefix));
    }

    return arr;
  }

  public getAddressFromPuzzleHash(puzzleHash: string, prefix: string): string {
    const address = bech32m.encode(prefix, bech32m.toWords(utility.fromHexString(puzzleHash)));
    return address;
  }

  public getPuzzleHashFromAddress(address: string): string {
    const hex = Bytes.from(bech32m.decodeToBytes(address).bytes).hex();
    if (hex.length == 66 && hex[0] == "0" && hex[1] == "0") return hex.substr(2);
    return hex;
  }

  public async getPuzzleHashes(privateKey: Uint8Array, prefix: string, startIndex = 0, endIndex = 10): Promise<string[]> {
    return (await this.getPuzzleDetails(privateKey, prefix, startIndex, endIndex)).map(_ => _.hash);
  }

  public async getPuzzleDetails(privateKey: Uint8Array, prefix: string, startIndex = 0, endIndex = 10): Promise<PuzzleDetail[]> {
    return await this.getPuzzleDetailsInner(privateKey, async (spk) => this.getPuzzle(spk), startIndex, endIndex, prefix, true)
  }

  public async getCatPuzzleDetails(privateKey: Uint8Array, assetId: string, prefix: string, startIndex = 0, endIndex = 10): Promise<PuzzleDetail[]> {
    return await this.getPuzzleDetailsInner(privateKey, async (spk) => this.getCatPuzzle(spk, assetId), startIndex, endIndex, prefix, true)
  }

  private async getPuzzleDetailsInner(
    privateKey: Uint8Array,
    getPuzzle: (pubkey: string) => Promise<string>,
    startIndex: number,
    endIndex: number,
    prefix: string,
    includeUnhardened = false): Promise<PuzzleDetail[]> {
    const derive = await utility.derive(privateKey, true);
    const deriveUnhardened = await utility.derive(privateKey, false);
    const details: PuzzleDetail[] = [];
    const add = async (privkey: PrivateKey, hardened: boolean) => {
      const pubkey = utility.toHexString(privkey.get_g1().serialize());
      const synpubkey = await this.getSyntheticKey(pubkey);
      const puzzle = await getPuzzle(synpubkey);
      const hash = await this.getPuzzleHashFromPuzzle(puzzle);
      const address = this.getAddressFromPuzzleHash(hash, prefix);
      details.push({ privateKey: privkey, hash, address, puzzle, type: hardened ? "Hardened" : "Observed" });
    }
    for (let i = startIndex; i < endIndex; i++) {
      const privkey = derive([12381, 8444, 2, i]);
      await add(privkey, true);
      if (includeUnhardened) {
        const privkey = deriveUnhardened([12381, 8444, 2, i]);
        await add(privkey, false);
      }
    }

    return details;
  }

  public getPrivateKeyFromHex(sk_hex: string): PrivateKey {
    const privateKey = utility.fromHexString(sk_hex);
    const BLS = Instance.BLS;
    if (!BLS) throw new Error("BLS not initialized");
    const sk = BLS.PrivateKey.from_bytes(privateKey, false);
    return sk;
  }

  public async getCatPuzzleHashes(privateKey: Uint8Array, assetId: string, prefix: string, startIndex = 0, endIndex = 10): Promise<string[]> {
    return (await this.getCatPuzzleDetails(privateKey, assetId, prefix, startIndex, endIndex)).map(_ => _.hash);
  }

  public async calcPuzzleResult(puzzle_reveal: string, solution: string, ...args: string[]): Promise<string> {
    let output: string[] = [];
    clvm_tools.setPrintFunction((...args) => output = args)

    clvm_tools.go("brun", puzzle_reveal, solution, "--experiment-backend", "rust", ...args);
    const result = output[0];

    const modname = modsdict[puzzle_reveal];
    if (result.startsWith("FAIL"))
      throw new Error(`Error calculating puzzle [${modname ?? puzzle_reveal.slice(0, 200)}] from solution [${solution.slice(0, 200)}]: ${result.slice(0, 200)}`);

    return result;
  }

  public async compileRun(puzzle_source: string): Promise<string> {
    let output: string[] = [];
    clvm_tools.setPrintFunction((...args) => output = args)

    clvm_tools.go("run", puzzle_source);
    const result = output[0];

    if (result.startsWith("FAIL")) throw new Error(result);

    return result;
  }

  public parseConditions(conditonString: string): ConditionEntity[] {
    try {
      const conds = Array.from(assemble(conditonString).as_iter())
        .map(cond => ({
          code: cond.first().atom?.at(0) ?? 0,
          args: Array.from(cond.rest().as_iter()).map(_ => _.listp() ? Array.from(_.as_iter()).map(_ => _.atom?.raw()) : _.atom?.raw())
        }));

      return conds;
    }
    catch (err) {
      console.warn("failed to parse condition: " + conditonString);
      throw err;
    }
  }

  public getSettlementPaymentsPuzzle(): string {
    const puzzle = `(a (q 2 10 (c 2 (c 3 ()))) (c (q (51 . 62) (a (i 5 (q 4 (c 12 (c (a 30 (c 2 (c 9 ()))) ())) (a 22 (c 2 (c 25 (c (a 10 (c 2 (c 13 ()))) ()))))) ()) 1) (a (i 5 (q 4 (c 8 9) (a 22 (c 2 (c 13 (c 11 ()))))) (q . 11)) 1) 2 (i (l 5) (q 11 (q . 2) (a 30 (c 2 (c 9 ()))) (a 30 (c 2 (c 13 ())))) (q 11 (q . 1) 5)) 1) 1))`;
    return puzzle;
  }

  public getEmptyPrivateKey(
  ): PrivateKey {
    const BLS = Instance.BLS;
    if (!BLS) throw new Error("BLS not initialized");
    return BLS.PrivateKey.from_bytes(new Uint8Array(32), false);
  }

  async executePuzzle(puz: string, solution: string): Promise<ExecuteResult> {
    const solution_result = await this.calcPuzzleResult(puz, solution);

    const conds = assemble(solution_result);
    const argarr = Array.from(conds.as_iter()).map((_) => Array.from(_.as_iter()).map((_) => disassemble(_)));
    const solution_results = argarr.map((_) => ({ op: Number(_[0]), args: _.slice(1) }));
    return { raw: solution_result, conditions: solution_results };
  }
}

export default new PuzzleMaker();
