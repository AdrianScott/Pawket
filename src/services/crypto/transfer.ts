import { PrivateKey, G1Element, ModuleInstance } from "@chiamine/bls-signatures";
import { Bytes, bigint_from_bytes, bigint_to_bytes } from "clvm";
import { OriginCoin, PuzzleInfo, SpendBundle } from "@/models/wallet";
import store from "@/store";
import puzzle from "./puzzle";

class CoinConditions {
  public static CREATE_COIN(puzzlehash: string, amount: bigint): string[] {
    return ["51", this.prefix0x(puzzlehash), this.prefix0x(Bytes.from(bigint_to_bytes(amount, { signed: true })).hex())];
  }
  public static CREATE_COIN_ANNOUNCEMENT(message: string): string[] {
    return ["60", message];
  }
  public static ASSERT_COIN_ANNOUNCEMENT(announcementId: string): string[] {
    return ["61", announcementId];
  }

  public static prefix0x(str: string): string {
    return str.startsWith("0x") ? str : "0x" + str;
  }
}

class Transfer {
  static readonly GROUP_ORDER = BigInt(bigint_from_bytes(Bytes.from("0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFFFFF00000001", "hex")));
  static readonly DEFAULT_HIDDEN_PUZZLE_HASH = Bytes.from("711d6c4e32c92e53179b199484cf8c897542bc57f2b22582799f9d657eec4699", "hex");
  static readonly AGG_SIG_ME_ADDITIONAL_DATA = Bytes.from("ccd5bb71183532bff220ba46c268991a3ff07eb358e8255a65c30a2dce0e5fbb", "hex");

  public async generateSpendBundle(
    coins: OriginCoin[],
    sk_hex: string,
    tgt_address: string,
    amount: bigint,
    fee: bigint,
    change_address: string): Promise<SpendBundle | null> {
    //   const sig = await this.generateSignature(coins, sk_hex,)
    // }

    // public async generateSignature(coins: CoinItem[], sk_hex: string, tgt_address: string, amount: number, fee: number, change_address: string) {
    if (!store.state.app.bls) return null;
    const BLS = store.state.app.bls;
    if (coins.length < 1) return null;
    const coin = this.findPossibleSmallest(coins, amount + fee);
    if (!coin) return null;
    if (amount + fee > coin.amount) return null;

    const gen_sk = BLS.PrivateKey.from_bytes(Bytes.from(sk_hex, "hex").raw(), true);
    const puzzles = await puzzle.getPuzzles(gen_sk.serialize(), 0, 10);
    const puzzleDict: { [key: string]: PuzzleInfo } = Object.assign({}, ...puzzles.map((x) => ({ [CoinConditions.prefix0x(x.puzzleHash)]: x })));
    const puz = puzzleDict[coin.puzzle_hash];
    if (!puz) return null;
    const puzzle_reveal = CoinConditions.prefix0x(await puzzle.encodePuzzle(puz.puzzle));
    const sk = BLS.PrivateKey.from_bytes(puz.privateKey, true);

    const tgt_hex = puzzle.getPuzzleHashFromAddress(tgt_address);
    const change_hex = puzzle.getPuzzleHashFromAddress(change_address);
    const a = bigint_to_bytes(BigInt(coin.amount), { signed: true });
    const pci = Bytes.from(coin.parent_coin_info, "hex");
    const ph = Bytes.from(coin.puzzle_hash, "hex");
    const cont = pci.concat(ph).concat(a);
    const coinname = Bytes.SHA256(cont);
    // const hash = Bytes.SHA256(cont).hex();
    // console.log(hash);

    // const pk = Bytes.from(sk.get_g1().serialize()).hex();
    // console.log("sk", sk, "pk", pk)
    const synthetic_sk = this.calculate_synthetic_secret_key(BLS, sk, Transfer.DEFAULT_HIDDEN_PUZZLE_HASH.raw());
    // console.log("synthetic_sk", Bytes.from(synthetic_sk.serialize()).hex())

    const delegated_puzzle_solution = this.getDelegatedPuzzle([
      CoinConditions.CREATE_COIN(tgt_hex, amount),
      CoinConditions.CREATE_COIN(change_hex, coin.amount - amount - fee),
    ]);
    const delegated_puzzle_solution_treehash = await puzzle.getPuzzleHashFromPuzzle(delegated_puzzle_solution);
    const solution_reveal = "(() " + delegated_puzzle_solution + " ())";
    const solution = CoinConditions.prefix0x(await puzzle.encodePuzzle(solution_reveal));
    // console.log(delegated_puzzle_solution_treehash);


    const message = Uint8Array.from([...Bytes.from(delegated_puzzle_solution_treehash, "hex").raw(), ...coinname.raw(), ...Transfer.AGG_SIG_ME_ADDITIONAL_DATA.raw()]);
    // console.log(synthetic_sk, delegated_puzzle_solution_treehash, coinname, AGG_SIG_ME_ADDITIONAL_DATA,  message);
    // console.log("message hex", Bytes.from(message).hex());

    const signature = BLS.AugSchemeMPL.sign(synthetic_sk, message);
    // console.log(Bytes.from(signature.serialize()).hex());

    const sig = Bytes.from(signature.serialize()).hex();

    return {
      aggregated_signature: CoinConditions.prefix0x(sig),
      coin_spends: [{ coin, puzzle_reveal, solution }]
    }
  }

  private findPossibleSmallest(coins: OriginCoin[], num: bigint): OriginCoin | null {
    const sortcoins = coins.sort((a, b) => Number(a.amount - b.amount));
    for (let i = 0; i < sortcoins.length; i++) {
      const coin = sortcoins[i];
      if (coin.amount >= num) return coin;
    }

    return null;
  }

  public getDelegatedPuzzle(conditions: string[][]): string {
    return "(q " + conditions.map(_ => "(" + _.join(" ") + ")").join(" ") + ")";
  }

  private calculate_synthetic_offset(public_key: G1Element, hidden_puzzle_hash: Uint8Array): bigint {
    const blob = Bytes.SHA256(new Uint8Array([...public_key.serialize(), ...hidden_puzzle_hash]));
    let offset = bigint_from_bytes(blob, { signed: true })
    while (offset < 0) offset += Transfer.GROUP_ORDER;
    offset %= Transfer.GROUP_ORDER;
    return offset;
  }

  private calculate_synthetic_secret_key(BLS: ModuleInstance, secret_key: PrivateKey, hidden_puzzle_hash: Uint8Array): PrivateKey {
    const secret_exponent = bigint_from_bytes(Bytes.from(secret_key.serialize()), { signed: true });
    const public_key = secret_key.get_g1();
    const synthetic_offset = this.calculate_synthetic_offset(public_key, hidden_puzzle_hash);
    const synthetic_secret_exponent = (secret_exponent + synthetic_offset) % Transfer.GROUP_ORDER
    const blob = bigint_to_bytes(synthetic_secret_exponent).raw();
    const synthetic_secret_key = BLS.PrivateKey.from_bytes(blob, true)
    return synthetic_secret_key;
  }
}

export default new Transfer();
