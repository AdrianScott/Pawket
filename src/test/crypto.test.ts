import puzzle from "@/services/crypto/puzzle";
import utility from "@/services/crypto/utility";
import transfer from "@/services/transfer/transfer";
import { getCoinName } from "@/services/coin/coinUtility";
import { Instance } from "@/services/util/instance";

beforeAll(async () => {
  await Instance.init();
})

test('Basic Cryptography', async () => {
  const hash = await utility.hash("test");
  expect(hash).toMatchSnapshot("hash");

  const privkey = utility.fromHexString("67b3dcf5ba985f77b7bb78b3edfd7e501f4669a3530b74f2247256e38b0529e2");
  const sk = await utility.getPrivateKey(privkey);
  const masterprikey = utility.toHexString(sk.serialize());
  const masterpubkey = utility.toHexString(sk.get_g1().serialize());
  expect(masterprikey).toMatchSnapshot("masterprikey");
  expect(masterpubkey).toMatchSnapshot("masterpubkey");
  expect(sk.get_g1().get_fingerprint().toString()).toMatchSnapshot("fingerprint");

  const derive = await utility.derive(privkey, true);
  const farmerpubkey = utility.toHexString(derive([12381, 8444, 0, 0]).get_g1().serialize());
  expect(farmerpubkey).toMatchSnapshot("farmerpubkey");
  const poolpubkey = utility.toHexString(derive([12381, 8444, 1, 0]).get_g1().serialize());
  expect(poolpubkey).toMatchSnapshot("poolpubkey");
  const walletprikey = utility.toHexString(derive([12381, 8444, 2, 0]).serialize());
  expect(walletprikey).toMatchSnapshot("walletprikey");
  const walletpubkey = utility.toHexString(derive([12381, 8444, 2, 0]).get_g1().serialize());
  expect(walletpubkey).toMatchSnapshot("walletpubkey");

  const adr = await puzzle.getAddress(walletpubkey, "xch");
  expect(adr).toMatchSnapshot("address");
});

test('Puzzle Assemble', async () => {
  const gt1 = transfer.getDelegatedPuzzle([
    ["a", "b", "c"],
    ["d", "e", "f"],
  ]);
  expect(gt1).toMatchSnapshot();

  const gt2 = transfer.getDelegatedPuzzle([
    ["a", "b", "c"],
    ["d", "e", ["f", `"g"`], []],
  ]);
  expect(gt2).toMatchSnapshot();
});

test('Coin Name', async () => {
  const coinname = getCoinName({
    amount: 1000000000n,
    parent_coin_info: "0xcd299604b459e5ff20da17627d684ea143fc1b5b4165166943729d2d24305de8",
    puzzle_hash: "0x484aaab0cda1b0149df9adeddb0a5d28976ad259ff6173f4488cbc68f095eb79",
  });
  expect(coinname).toMatchSnapshot("coinname");
});