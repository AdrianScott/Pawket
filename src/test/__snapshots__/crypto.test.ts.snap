// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Cryptography: address 1`] = `"xch13akv0y3er0qvdjwzks2gm4ljj7qpynrh6rcsnwc6y0hyfgzdj89sr43zcp"`;

exports[`Basic Cryptography: farmerpubkey 1`] = `"99f181b1b30865339eea4e61172e64282ed1793022363f6635affbfb2aa2f4384ef10df47a445f92d685c7265488777e"`;

exports[`Basic Cryptography: fingerprint 1`] = `"2308828858"`;

exports[`Basic Cryptography: hash 1`] = `"9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"`;

exports[`Basic Cryptography: masterprikey 1`] = `"67b3dcf5ba985f77b7bb78b3edfd7e501f4669a3530b74f2247256e38b0529e2"`;

exports[`Basic Cryptography: masterpubkey 1`] = `"89d4ca795881dc192cf7e0bcf22528c3d7708f57aa644647ae2506cd10405dc528b36e450ab901675f7a6289a4335a22"`;

exports[`Basic Cryptography: poolpubkey 1`] = `"9896eb01246db07360b4caba1861a6d13c5be923e28a68ba8429d841c5dfa429e05ef639003793c4715a8be4578abe9e"`;

exports[`Basic Cryptography: walletprikey 1`] = `"0408d4c5e7c97af49dd5605db1907a231761c0cba816c778e74c8de24d0793cd"`;

exports[`Basic Cryptography: walletpubkey 1`] = `"8b58921998b2337fd9d4a410e8bd11dfda435feb92dc3d6f3e111194b99a0391f39cf9b9534a44d154c9d7492400f36b"`;

exports[`Coin Name: coinname 1`] = `"c62152493a6f3fbca7a918a8258e52f84c4e8aa286b8ebf1b0d5e5dda7fa7e6f"`;

exports[`Puzzle Assemble 1`] = `"(q (a b c) (d e f))"`;

exports[`Puzzle Assemble 2`] = `"(q (a b c) (d e (f \\"g\\") ()))"`;
